build_support/generate_modelsmodule.py

```diff
@@ -0,0 +1,340 @@
+    argument, as each file could contain multiple model definitions.
+
+    This function does not return anything, but instead sets the
+    global variables `includes` and `models` to be used by the code
+    generation function.
+
+    """
+
+    global includes, models
+
+    includes = {}
+    models = {}
+
+    for model_file in model_names:
+        guards, model_types_names = get_models_from_file(model_file)
+        for tp, nm in model_types_names:
+            # Assemble a nested dictionary for the includes:
+            fname = model_file + ".h"
+            if tp in includes:
+                if guards in includes[tp]:
+                    includes[tp][guards].add(fname)
+                else:
+                    includes[tp][guards] = set([fname])
+            else:
+                includes[tp] = {guards: set([fname])}
+
+            if (Path(srcdir) / "models" / f"{model_file}_impl.h").is_file():
+                includes[tp][guards].add(f"{model_file}_impl.h")
+
+            # Assemble a nested dictionary for the models:
+            if tp in models:
+                if guards in models[tp]:
+                    models[tp][guards].append(nm)
+                else:
+                    models[tp][guards] = [nm]
+            else:
+                models[tp] = {guards: [nm]}
+
+
+def start_guard(guards):
+    """Print an #ifdef line with preprocessor guards if needed."""
+
+    if guards:
+        guard_str = " && ".join([f"defined( {guard} )" for guard in guards])
+        return f"#if {guard_str}\n"
+    else:
+        return ""
+
+
+def end_guard(guards):
+    """Print an #endif line for the preprocessor guards if needed."""
+    return "#endif\n" if guards else ""
+
+
+def generate_modelsmodule():
+    """Write the modelsmodule implementation out to file.
+
+    This is a very straightforward function that prints several blocks
+    of C++ code to the file modelsmodule.cpp in the `blddir` handed as
+    a commandline argument to the script. The blocks in particular are
+
+    1. the copyright header.
+    2. a list of generic NEST includes
+    3. the list of includes for the models to build into NEST
+    4. some boilerplate function implementations needed to fulfill the
+       Module interface
+    5. the list of model registration lines for the models to build
+       into NEST
+
+    The code is enriched by structured C++ comments as to make
+    debugging of the code generation process easier in case of errors.
+
+    """
+
+    fname = Path(srcdir) / "doc" / "copyright_header.cpp"
+    with open(fname, "r") as file:
+        copyright_header = file.read()
+
+    fname = "modelsmodule.cpp"
+    modeldir = Path(blddir) / "models"
+    modeldir.mkdir(parents=True, exist_ok=True)
+    with open(modeldir / fname, "w") as file:
+        file.write(copyright_header.replace("{{file_name}}", fname))
+        file.write(
+            dedent(
+                """
+            #include "modelsmodule.h"
```

Comment:
Is it better to generate this file via a jinja template?

Reply:
I also considered this, but since this is part of the run of `cmake` and should work everywhere, I thought it would be better to avoid external dependencies.